eqcorrscan/tests/correlate_test.py

```diff
@@ -2,108 +2,399 @@
+        assert hasattr(corr.XCOR_FUNCS['func_test'], 'registered')
+
+
+class TestRegisterNormXcorrs:
+    """ Tests for register_normxcorr function, which holds global context
+    for which xcorr to use """
+
+    # helper functions
+    def name_func_is_registered(self, func_name):
+        """ return True if func is registered as a normxcorr func """
+        name = func_name.__name__ if callable(func_name) else func_name
+        return name in eqcorrscan.utils.correlate.XCOR_FUNCS
+
+    # tests
+    def test_register_as_decorator_no_args(self):
+        """ ensure register_normxcorr works as a decorator with no args """
+
+        @register_array_xcorr
+        def func1(templates, stream, pads, *args, **kwargs):
+            pass
+
+        assert self.name_func_is_registered(func1)
+
+    def test_register_as_decorator_with_args(self):
+        """ ensure register can be used as a decorator with args """
+
+        @register_array_xcorr(name='func2')
+        def func(templates, stream, pads, *args, **kwargs):
+            pass
+
+        assert self.name_func_is_registered('func2')
+
+    def test_register_as_callable(self):
+        """ ensure register can be used as a callable to take a name
+        and a normxcorr func """
+        func = gen_xcorr_func('funky')
+        register_array_xcorr(name='func3', func=func)
+        assert self.name_func_is_registered('func3')
+
+    def test_set_default(self):
+        """ ensure the default can be overwritten """
+        func = gen_xcorr_func('funky')
+        corr.register_array_xcorr(func, is_default=True)
+        assert corr.XCOR_FUNCS['default'] is func
+
+    def test_register_bad_func_rasies(self):
+        """ test trying to register a non-supported function raises """
+        func = corr.XCOR_FUNCS['default']
+
+        with pytest.raises(ValueError):
+            @func.register('not_supported_value')
+            def func():
+                pass
+
+
+class TestRegisterAlternativeConcurrency:
+    """ Tests for registering alternative concurrency functions """
+    counter = defaultdict(lambda: 0)
+
+    # helper functions
+    def new_multi(self, template_dict, stream_dict, pad_dict, seed_ids):
+        pass
+
+    # fixtures
+    @pytest.fixture
+    def r_normxcorr(self):
+        """ return a registered normxcorr function """
+        return register_array_xcorr(gen_xcorr_func('normxcorr'))
+
+    @pytest.fixture
+    def normxcorr_new_multithread(self, r_normxcorr):
+        """ register the new multithread method """
+        func = measure_counts(self, self.new_multi)
+        r_normxcorr.register('multithread')(func)
+        r_normxcorr.multithread(None, None, None, None)
+        yield func.__name__
+        self.counter.pop(func.__name__)
+
+    # tests
+    def test_new_method_was_called(self, normxcorr_new_multithread):
+        """ ensure the new method was called """
+        assert self.counter[normxcorr_new_multithread]
 
 
 if __name__ == '__main__':
```

Comment:
Note to selves - because these are all pytest now, this `if __name__ == '__main__':` doesn't make sense/do anything - @d-chambers is there a py.test alternative?

Reply:
From what I read [here](https://stackoverflow.com/questions/20992731/meaning-of-unittest-main-in-python-unittest-module), I don't think we need it. When calling `pytest` these tests will get auto-discovered and run. If you want to just run the tests in this file you can specify that when invoking pytest like so:
```bash
pytest eqcorrscan/tests/correlate_test.py
```
You can use pytest.main() as described [here](https://docs.pytest.org/en/latest/usage.html#calling-pytest-from-python-code) if you want to be able to run the tests from python rather than using the CLI. Is that not the same thing as `unittest.main()`?

Reply:
So pytest.main doesn't do the same as unittest.main - I think we should just remove the whole `if __name__ == __main__` statement in here - calling via pytest is cleaner anyways.